src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -641,3 +649,295 @@ ExtractPartitionValue(Query *query, Var *partitionColumn)
+
+	if (partitionMethod != DISTRIBUTE_BY_HASH)
+	{
+		return false;
+	}
+
+	/* extract range table entries */
+	ExtractRangeTableEntryWalker((Node *) query, &rangeTableList);
+
+	/* query can have only one range table of type RTE_RELATION */
+	if (list_length(rangeTableList) != 1)
+	{
+		return false;
+	}
+
+	rangeTableEntry = (RangeTblEntry *) linitial(rangeTableList);
+	if (rangeTableEntry->rtekind != RTE_RELATION)
+	{
+		return false;
+	}
+
+#if (PG_VERSION_NUM >= 90500)
+	if (rangeTableEntry->tablesample)
+	{
+		return false;
+	}
+#endif
+
+	if (joinTree == NULL)
+	{
+		return false;
+	}
+
+	quals = joinTree->quals;
+	if (quals == NULL)
+	{
+		return false;
+	}
+
+	/* convert list of expressions into expression tree */
+	if (quals != NULL && IsA(quals, List))
+	{
+		quals = (Node *) make_ands_explicit((List *) quals);
+	}
+
+	/* partition column must be used in a simple equality match check */
+	partitionColumnMatchExpression =
+		ContainsPartitionColumnMatchExpression(quals, partitionColumn);
+
+	if (!partitionColumnMatchExpression)
+	{
+		return false;
+	}
+
+	/* make sure partition column is used only once in the query */
+	varClauseList = pull_var_clause_default(quals);
+	foreach(varClauseCell, varClauseList)
+	{
+		Var *column = (Var *) lfirst(varClauseCell);
+		if (equal(column, partitionColumn))
+		{
+			partitionColumnReferenceCount++;
+		}
+	}
+
+	if (partitionColumnReferenceCount != 1)
+	{
+		return false;
+	}
+
+	/*
+	 * We need to make sure there is at least one shard for this
+	 * hash partitioned query.
+	 */
+	shardCount = ShardIntervalCount(distributedTableId);
+	if (shardCount == 0)
+	{
+		return false;
+	}
+
+	/*
+	 * Cursor queries are not router plannable. Opening a cursor creates a transaction,
+	 * we disallow cursor queries by rejecting queries inside transactions. These
+	 * queries can be served by regular planner/executor.
+	 */
+	inTransaction = IsInTransactionChain(true);
+	if (inTransaction)
+	{
+		return false;
+	}
+	return true;
+}
+
+
+/*
+ * ContainsPartitionColumnMatchExpression returns true if the query contains an exact
+ * match (equal) expression on the partition column. The function returns true only
+ * if the match expression has an AND relation with the rest of the expression tree.
+ */
+static bool
+ContainsPartitionColumnMatchExpression(Node *node, Var *partitionColumn)
+{
+	if (node == NULL)
+	{
+		return false;
+	}
+
+	if (IsA(node, OpExpr))
+	{
+		OpExpr *opExpr = (OpExpr *) node;
+		bool simpleExpression = SimpleOpExpression((Expr *) opExpr);
+		bool partitionColumnInExpr = false;
+		char *operatorName = NULL;
+		int operatorNameComparison = 0;
+		bool usingEqualityOperator = false;
+
+		if (!simpleExpression)
+		{
+			return false;
+		}
+
+		partitionColumnInExpr = OpExpressionContainsColumn(opExpr, partitionColumn);
+		if (!partitionColumnInExpr)
+		{
+			return false;
+		}
+
+		operatorName = get_opname(opExpr->opno);
+		operatorNameComparison = strncmp(operatorName, EQUAL_OPERATOR_STRING,
```

Comment:
@anarazel So, we're going to merge this today. If you think that this is a critical issue, could you open an item for that (or note here so that we can open an issue)?

Considering the e-mail titled `checking if expression is simple equality check`, we think that checking the operator via it's oid is something nice to have, but, that's not a must have since there are some codes in postgres which does similar thing. Also, given that we have the same pattern used in 2 other places, it'd be nice to fix them as well if that's something critical.